### DISCUSSION TITLE
Bundlesize of ariakit is huge

### DISCUSSION THREAD

**Discussion by artalar** [ariakit/ariakit#1705] [Questions] [answered]
[Here what I get](https://bundle.js.org/?q=ariakit%2Fform&config=%7B%22analysis%22%3A%22treemap%22%7D) for `ariakit/form`: 266.22KB -> 85.72KB (gzip). Pretty huge!

<img width="916" alt="image" src="https://user-images.githubusercontent.com/27290320/182765494-5343a029-f660-40b1-b584-557cd392283a.png">

<img width="642" alt="image" src="https://user-images.githubusercontent.com/27290320/182765353-19cdbcaf-cbb9-48d5-bd14-b82c847478bd.png">

`@floating-ui/core` took two times more, than clean build from npm

<img width="391" alt="image" src="https://user-images.githubusercontent.com/27290320/182765831-dab8d0bf-fd6e-4f8e-b9ef-8085c91e1156.png">

<img width="488" alt="image" src="https://user-images.githubusercontent.com/27290320/182765915-512549db-ec99-4bf5-8067-ac2fb902a516.png">

May be, build output target is too low? Try this `"browserslist": [ "last 4 chrome versions" ]`, what the difference will be?

**Comment by diegohaz** [comment, marked as answer]
bundlejs is producing incorrect results. I recommend checking it in a real app.

All `ariakit/form` totalize around 18.5 kB gzip, and most people wouldn't even use all its modules. The following imports, for example, totalize 12.6 kB:

```js
import {
  Form,
  FormError,
  FormInput,
  FormLabel,
  FormSubmit,
  useFormState,
} from "ariakit/form";
```

The screenshots you shared include the whole `ariakit` library (not only `ariakit/form`), [which is around 39 kB gzip](https://bundlephobia.com/package/ariakit@2.0.0-next.37), **and** react-dom, which is a *peer* dependency, [with 42 kB gzip](https://bundlephobia.com/package/react-dom@18.2.0).

**Reply by okikio** [reply]
@diegohaz bundlejs bundles all packages including react and react-dom, in this case it seems you'd like have react and react-dom not bundled, so you can mark them as external to remove them from the final size [https://bundlejs.com/?q=https://unpkg.com/ariakit/esm/form/index.js&treeshake=[{Form,FormError,FormInput,FormLabel,FormSubmit,useFormState}]&config={"esbuild":{"external":["react","react-dom"]}}&bundle](https://bundlejs.com/?q=https%3A%2F%2Funpkg.com%2Fariakit%2Fesm%2Fform%2Findex.js&treeshake=%5B%7B+Form%2CFormError%2CFormInput%2CFormLabel%2CFormSubmit%2CuseFormState+%7D%5D&config=%7B%22esbuild%22%3A%7B%22external%22%3A%5B%22react%22%2C%22react-dom%22%5D%7D%7D&bundle)

Note: esbuild isn't the best at treeshaking...so it sometimes bundle functions that aren't in use. I tried switching to rollup a while back but the perf hit, isn't worth the small improvement in treeshaking that rollup provides

**Reply by diegohaz** [reply]
@okikio Thanks for the clarification. I suggest using that configuration by default, as it seems a lot of people get confused by that. I think it's a great tool, but I receive too many reports based on wrong Bundlejs results.

That said, it would be great to link to Bundlejs in the Ariakit docs of it produces correct results by default (I can use the link you shared as well).

**Reply by okikio** [reply]
Yeah, I'll probably do that. What other libraries do you know of that would be beneficial to mark as external?

**Comment by artalar** [comment]
@okikio check this out

**Reply by okikio** [reply]
Thanks for the shout out, that's a flaw in the way bundle analysis currently works where it's not treeshaking individual imports properly, I'm working on a re-write that gives better and more accurate results, right now the results it gives are the non-treeshaken sizes. 

Treeshaking Normal bundles work, but treeshaking individual modules as you find them, is a bit difficult, esbuild is fairly good at the first but it isn't as good at doing the latter .
